### ISSUE TITLE
Experiment: Qwen3.5-0.8B Tool-Jev fine-tuning on two DGX Sparks

### ISSUE THREAD

**Issue by OriNachum** [agentculture/nvsh#46]
## Goal

Test whether **Qwen3.5-0.8B** can become a small, fast **Tool-Jev** for nvsh: given the current command/error/context plus the available actions/tools, produce a calibrated decision about **what action/tool to take (or not take)**.

We now have two DGX Sparks available:

- `spark`
- `spark2`

Use them in parallel to explore **two different fine-tuning approaches** rather than serially optimizing one recipe.

## Why

A 0.8B model is small enough to target:

- Jetson Orin Nano
- cheap CPU/GPU K8s services
- always-resident local decision services

Qwen3.5-0.8B is Apache-2.0, which is important because the resulting model should remain easy to share externally (including with NVIDIA), subject to the licenses/provenance of the training data we use.

The experiment should answer a more specific question than “can a tiny model call tools?”:

> Can we train a small model to behave like a specialized Jev-style decision component for tool use, with useful confidence/probability information and strong abstention behavior?

## Parallel experiment

Run both tracks from the same base model and evaluate them on the same held-out set.

### Track A — `spark`: generative structured tool caller

Fine-tune Qwen3.5-0.8B using conventional SFT.

Input should include the minimum useful state, for example:

- user/agent intent
- failed command and exit code when relevant
- bounded stdout/stderr
- environment metadata
- candidate tools/actions and schemas
- constraints/policy

Target output should be a strict compact structure such as:

```json
{
  "action": "tool",
  "tool": "shell.exec",
  "arguments": {
    "command": "..."
  }
}
```

or:

```json
{
  "action": "abstain"
}
```

Focus on:

- exact tool selection
- exact argument extraction
- valid structured output
- strong `no_action` / abstain behavior
- very short generations
- no chain-of-thought requirement

### Track B — `spark2`: Jev-style decision / scoring variant

Explore a non-free-form formulation where the model primarily **scores candidate actions** rather than writes an answer.

Possible implementation directions:

1. candidate action labels expressed as constrained tokens and scored from logits;
2. classification head over the final/pooled latent;
3. pairwise scoring: `state + candidate action -> score`;
4. another compact formulation if it preserves the key property: **decision + confidence without requiring prose generation**.

Example conceptual output:

```text
shell.exec        0.91
file.read         0.05
ask_user          0.02
no_action         0.02
```

The probabilities do not need to be perfectly calibrated initially, but the experiment should measure calibration rather than assume softmax values are probabilities.

## Shared dataset

Both tracks should use the **same underlying examples and split** so we can compare the training objective rather than different data.

Include:

- correct tool calls
- argument extraction
- commands that require no intervention
- ambiguous cases where abstaining is correct
- unsafe/invalid candidates
- overlapping tools where selecting the right tool is non-trivial
- negative examples where a plausible tool exists but should not be called
- examples derived from real nvsh failure/recovery flows where possible

Keep training-data provenance explicit. Prefer data we generate/control or datasets whose licenses allow redistribution of the resulting work.

## Context-size experiment

Do not optimize for maximum model context.

Start with a deliberately small operational budget:

- primary target: **2K**
- comparison: **4K**
- only test 8K+ if examples demonstrate a real need

Tool-Jev should receive a compact decision state rather than the entire terminal/session history.

Track peak memory and latency for each context setting.

## Evaluation

Create one held-out evaluation suite used by both tracks.

Measure at minimum:

- tool-selection accuracy
- argument exact-match / schema validity
- abstention precision and recall
- false-positive tool invocation rate
- unsafe/invalid action rate
- confidence calibration (ECE/Brier score or equivalent)
- latency to first decision
- total decision latency
- tokens generated per decision
- VRAM/unified-memory usage
- model artifact size

Also evaluate behavior when the correct tool is **not** among the candidates.

## Deployment check

After training, test the better checkpoints with an inference configuration representative of the intended targets:

1. DGX Spark — reference
2. Jetson Orin Nano — target edge deployment if available
3. K8s-friendly serving profile — quantify memory/CPU/GPU requirements

For the initial FT experiment, both Sparks should work independently in parallel; do not require distributed training across the two machines.

## Deliverables

- [ ] Shared dataset + fixed train/validation/test split
- [ ] Baseline Qwen3.5-0.8B evaluation before FT
- [ ] Track A checkpoint + training recipe from `spark`
- [ ] Track B checkpoint/head + training recipe from `spark2`
- [ ] Shared evaluation script/harness
- [ ] Comparison table for accuracy, abstention, calibration, latency and memory
- [ ] 2K vs 4K context comparison
- [ ] Quantized inference test
- [ ] Recommendation for the next Tool-Jev iteration
- [ ] Document base-model and dataset licenses/provenance so the resulting model can be shared safely

## Success criteria

The experiment is successful if at least one fine-tuned variant:

1. materially beats the untuned Qwen3.5-0.8B baseline on tool selection;
2. reliably abstains instead of inventing calls;
3. produces usable confidence/scoring information;
4. is small/fast enough to make an always-resident Tool-Jev practical;
5. retains a clean enough license/provenance story to share the result externally, including with NVIDIA.

The point of using two Sparks is **breadth, not duplicated work**: run the two formulations concurrently, then converge based on measured results.

**Comment by OriNachum** [comment]
## Baseline from issue 39 (LFM2.5-350M): what carries over, what doesn't

Branch `feat/lfm-finetune-issue-39` ran the full recipe end to end on `spark`, and wrote it up in `docs/lfm-finetune.md` (the guide plus a dated run log). It is a good starting point for **Track A**. It does not cover **Track B**: there is no scoring head, no logit-scored candidates and no calibration metric.

**Issue 39's results, for scale.** Stock LFM2.5-350M against the tuned run r8 on the test side:

| | Stock | r8 |
|---|---|---|
| Right proposals | 0 of 32 | 28 of 32 |
| Escalations (this issue's "abstain") | 0 of 15 | 13 of 15 |
| Explained in words | 15 of 17 | 17 of 17 |
| Warm median | 290 ms | 119 ms |
| Wrong mutating proposals | 0 | 1 (bar requires 0, so not adopted) |

On NVIDIA's 104 Jetson skill evals the same recipe went from 35 to 52 (not-named 12 to 33 of 70).

### Reusable as is (both tracks)

- **Shared split:** `scripts/lfm-finetune/split.py`, seeded 70/15/15 train/val/test, stratified by answer, refuses the held-out file. This covers the "shared dataset + fixed split" deliverable. Reuse seed 39 and the same files for a like-for-like comparison.
- **Data:**
  - nvsh's corpus (`nvsh/tiers/corpus/dev.json`: 212 operation, 103 escalate, 116 explain entries);
  - reviewed variations (`augment.py`: one model generates, one copyedits, two reviewers must both accept, and deterministic guards reject identifier or answer leaks);
  - a train-only supplement of stop/restart contrasts.

  These are the "negative examples", "abstain" and "overlapping tools" cases this issue asks for, with provenance recorded per record.
- **Skills task:** `jetson_skills.py` provides 38 tools and NVIDIA's 104 evals as a never-trained test set, with a contamination scan. It serves as an independent tool-selection benchmark, including overlapping tools.
- **Measurement:** `measure.py` measures Tier 2 through the real launcher. It reports right proposals, escalations, wrong mutating proposals and wrong arguments as separate counts, plus latency, container memory and background load. `measure_skills.py` covers skill routing. Both refuse the test side without `--final`.
- **Runner and serving:** `pipeline.sh` (resumable stages, runnable outside an agent session), `stage_cache.py`, and `[tiers.lfm] hf_offline` (no token in the container).

### LFM-specific, must be redone for Qwen3.5-0.8B (not yet verified with Qwen)

1. **Chat template and tool-call format.** The guide's pitfalls (Pythonic calls, arguments must be an object, `operation` before `arguments`) are LFM2.5 template properties. Re-render one example with the real Qwen tokenizer before training.
2. **Loss mask.** `train.py` uses `apply_chat_template(..., return_assistant_tokens_mask=True)`, which needs generation markers in the template. Check that the mask is non-empty and covers only the answer.
3. **Thinking mode.** Training and serving must agree. This issue asks for "no chain-of-thought", so turn thinking off in both.
4. **vLLM tool parser.** `lfm2` is wrong for Qwen. Also check that the pinned vLLM image supports the model.
5. **Base and revision.** Set `BASE`/`BASE_REV` in the env file and measure stock at the same commit first.
6. **`release_bundle.py`** is hard-coded to the LFM Open License (licence check, NOTICE, and the model card's USD 10M threshold). Qwen needs its own licence handling; confirm the licence from the model repo.
7. **Hyperparameters** (20 epochs, lr 5e-4, LoRA r32/a64, batch 8) were tuned for the 350M. Re-tune on validation only. On the skills task, 8 epochs beat 20.
8. **Latency and memory** were measured at 350M on a GB10. Re-measure at 0.8B, and at 2K against 4K context as this issue plans (issue 39 ran at `ctx = 4096`).

### Caveats to plan around

- **The test side is no longer pristine.** Final runs have read it twice, and one entry was seen during a grep (lapse l2). It is still fair for a new model if nobody iterates on it.
- **The skills task has no validation side.** Three recipes were scored on the same 104 evals, so issue 39's pass is best-of-three (lapse l4). For Qwen, fix the recipe and state the margin before the first scored run, then score once. Better still, carve a validation slice first.
- **One output format for both tracks.** Track A's JSON (`{"action": "tool", ...}` / `{"action": "abstain"}`) differs from nvsh's propose/explain/escalate tools. Pick one mapping before training so both tracks and the stock baseline are scored on the same labels.
- **Calibration** (ECE/Brier) is not measured anywhere in issue 39's tooling. It is new work for both tracks.

- nvsh (Claude)
